File: src/core/ops/compose.ts
import type { Registry } from "../../registry/types.ts";
import { MenvError } from "../errors.ts";
import type { OpResult } from "./util.ts";
import { cloneRegistry, newPlan } from "./util.ts";

// Registry list management only — marker filling and .env.compose are Plan 3.
// On-disk existence of the file is the CLI layer's check (ops do no I/O).
export function planComposeBind(registry: Registry, input: { file: string }): OpResult {
  if (registry.compose.files.includes(input.file)) {
    throw new MenvError("VALIDATION", `"${input.file}" is already bound`);
  }
  const next = cloneRegistry(registry);
  next.compose.files.push(input.file);
  const plan = newPlan();
  plan.registry.push({ action: "set", path: "compose.files", summary: `bind compose file "${input.file}"` });
  return { next, plan };
}

export function planComposeUnbind(registry: Registry, input: { file: string }): OpResult {
  if (!registry.compose.files.includes(input.file)) {
    throw new MenvError("NOT_FOUND", `"${input.file}" is not bound (bound: ${registry.compose.files.join(", ") || "none"})`);
  }
  const next = cloneRegistry(registry);
  next.compose.files = next.compose.files.filter((f) => f !== input.file);
  const plan = newPlan();
  plan.registry.push({ action: "remove", path: "compose.files", summary: `unbind compose file "${input.file}"` });
  return { next, plan };
}

File: go/internal/core/ops/compose.go
package ops

import (
	"fmt"
	"strings"

	"github.com/nikrabaev/menv/internal/core"
	"github.com/nikrabaev/menv/internal/registry"
)

// PlanComposeBind registers a compose file in the registry.
func PlanComposeBind(r registry.Registry, input struct{ File string }) (OpResult, error) {
	for _, f := range r.Compose.Files {
		if f == input.File {
			return OpResult{}, &core.MenvError{
				Code:    core.ErrValidation,
				Message: fmt.Sprintf("%q is already bound", input.File),
			}
		}
	}
	next := CloneRegistry(r)
	next.Compose.Files = append(next.Compose.Files, input.File)
	plan := NewPlan()
	plan.Registry = append(plan.Registry, core.RegistryOp{
		Action:  "set",
		Path:    "compose.files",
		Summary: fmt.Sprintf("bind compose file %q", input.File),
	})
	return OpResult{Next: next, Plan: plan}, nil
}

// PlanComposeUnbind removes a compose file from the registry.
func PlanComposeUnbind(r registry.Registry, input struct{ File string }) (OpResult, error) {
	found := false
	for _, f := range r.Compose.Files {
		if f == input.File {
			found = true
			break
		}
	}
	if !found {
		bound := strings.Join(r.Compose.Files, ", ")
		if bound == "" {
			bound = "none"
		}
		return OpResult{}, &core.MenvError{
			Code:    core.ErrNotFound,
			Message: fmt.Sprintf("%q is not bound (bound: %s)", input.File, bound),
		}
	}
	next := CloneRegistry(r)
	filtered := make([]string, 0, len(next.Compose.Files)-1)
	for _, f := range next.Compose.Files {
		if f != input.File {
			filtered = append(filtered, f)
		}
	}
	next.Compose.Files = filtered
	plan := NewPlan()
	plan.Registry = append(plan.Registry, core.RegistryOp{
		Action:  "remove",
		Path:    "compose.files",
		Summary: fmt.Sprintf("unbind compose file %q", input.File),
	})
	return OpResult{Next: next, Plan: plan}, nil
}
